velox/docs/develop/expression-evaluation.rst

```diff
@@ -185,6 +185,35 @@ reordering during execution of the AND and OR expressions.
   :width: 600
   :align: center
 
+Flatten concat-like functions
+`````````````````````````````
+
+Functions that behave like associative operators can declare support for
+flattening. In that case, adjacent nodes of the same function are
+consolidated into one.
+
+A good example is concat(varchar,..) Presto function. Evaluating concat(a, b, c,
+d) is more efficient than evaluating concat(a, concat(b, concat(c, d))).
+Concatenating 4 columns at once allows to calculate the total amount of memory
+needed for the final result, allocate it in one chunk, then copy individual
+values to the right offsets. This saves on memory allocations and reduces data
+copy as compared to concatenating two columns at a time.
+
+Concat Presto function declares support for flattening allowing the expression
+compiler to convert concat(a, concat(b, concat(c, d))) expression to
+concat(a, b, c, d).
+
+Other functions that can leverage this optimization include concat(arrays,..) and
+map_concat(map,..).
+
+A function declaring support for flattening must have a signature with variadic
+arguments of the same type and matching return type.
```

Comment:
What do you mean by matching return type here? Does it need to be the same type? We only require "variadic arguments" to be the same type? If we don't do flatting, then returning type of a subexpression will be input type? Why there is subtle difference here? Thanks!

Reply:
Yes, both input and return types must match. Otherwise, it is not possible to chain the calls. f(x, f(y, z)) requires a, y, z and f(y, z) all have the same type.

Reply:
So match here actually means the same type? Thanks!